packages/core/src/ctype/CTypeLoader.ts

```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright (c) 2018-2023, BOTLabs GmbH.
+ *
+ * This source code is licensed under the BSD 4-Clause "Original" license
+ * found in the LICENSE file in the root directory of this source tree.
+ */
+
+import { ICType } from '@kiltprotocol/types'
+
+import { fetchFromChain } from './CType.chain.js'
+
+export type CTypeLoader = (id: ICType['$id']) => Promise<ICType>
+
+const loadCType: CTypeLoader = async (id) => {
+  return (await fetchFromChain(id)).cType
+}
+
+/**
+ * A factory for a CType loader that caches a CType definition once it has been loaded.
+ * Used in validating the credentialSubject of a [[KiltCredentialV1]] against the Claim Type referenced in its `type` field.
+ *
+ * @param initialCTypes An array of CTypes with which the cache is to be initialized.
+ * @returns A function that takes a CType id and looks up a CType definition in an internal cache, and if not found, tries to fetch it from the KILT blochchain.
+ */
+export function newCachingCTypeLoader(
+  initialCTypes: ICType[] = []
+): CTypeLoader {
+  const ctypes: Map<string, ICType> = new Map()
+
+  initialCTypes.forEach((ctype) => {
+    ctypes.set(ctype.$id, ctype)
+  })
+
+  async function getCType(id: ICType['$id']): Promise<ICType> {
```

Comment:
does that help in any way? is this to distinguish between say, network errors and not founds?

Reply:
I mean, you call. I guess a network error could throw, but a credential not found could return `null`. But that's my opinion. If you are fine with the function throwing in either case, I'm fine with that 😊

Reply:
Phew, that's a tricky one. Ultimately we're talking about the proper design of the CTypeLoader interface, which right now is `(id: ICType['$id']) => Promise<ICType>`. Making it a `(id: ICType['$id']) => Promise<ICType | null>` would mean you always have to handle two different error cases (reject or resolve to null), so I'd only introduce this if we can see how depending implementations (e.g. credential verification) would implement different logic/flows for each of these error cases. Otherwise we'd only introduce an annoyance I would say. Do you have an idea of how this would improve depending functions?

Reply:
I can only comment as a Rust programmer, that it would be much easier to have as much information as possible conveyed in the function signature. I agree probably handling two different error cases is an annoyance for a TS developer. It's just that exceptions in TS and JS require really good documentation, otherwise it's easy for a consumer to forget about a try-catch and the application to crash.

Reply:
so it seems that `fetchFromChain` also does not make a distinction between connection error and not found, so I'd leave the new CTypeLoader as it is right now as well, because the info we wanted to convey is simply missing